sdk/daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/crypto/Hash.scala

```diff
@@ -180,18 +212,248 @@ object Hash {
       val ss = set.toSeq.sorted[String]
       iterateOver(ss.iterator, ss.size)(_ addString _)
     }
+
+    final def addOptional[S](opt: Option[S], hashS: this.type => S => this.type): this.type = {
+      opt match {
+        case None => addInt(0)
+        case Some(value) => hashS(addInt(1))(value)
+      }
+    }
+  }
+
+  // Transaction Builder class with builder methods to recursively encode nodes
+  private sealed class TransactionBuilder(
+      protected val messageDigest: MessageDigestWithContext,
+      valueBuilder: ValueHashBuilder,
+  ) extends Builder(stringNumericToBytes, messageDigest.isContextAware) {
+
+    override protected def update(a: ByteBuffer, context: Option[String]): Unit = {
+      messageDigest.update(a, context)
+    }
+    override protected def update(a: Array[Byte], context: Option[String]): Unit = {
+      messageDigest.update(a, context)
+    }
+    override protected def doFinal(buf: Array[Byte]): Unit = {
+      assert(messageDigest.digest(buf, 0, underlyingHashLength) == underlyingHashLength)
+    }
+    def addContext(context: => String): this.type = {
+      withContext(context)(identity)
+    }
+    def withContext(context: => String)(f: this.type => this.type): this.type = {
+      contextO(context).map(c => s"# $c").foreach(messageDigest.addContext)
+      f(this)
+    }
+
+    private[lf] def addFromVersionNode(node: Node, nodes: Map[NodeId, Node]): NodeBuilder = {
+      NodeBuilder.builderForNode(node, messageDigest, valueBuilder).addNode(node, nodes)
+    }
+
+    @throws[NodeHashingError]
+    protected def addNodeFromNodeId(nodes: Map[NodeId, Node]): (this.type, NodeId) => this.type =
+      (builder, nodeId) => {
+        nodes.get(nodeId) match {
+          // Remove the node from the map once we've used it
+          // This serves as an additional check that nodes are only hashed once
+          case Some(node) =>
+            // discard the resulting builder and return this one because addFromVersionNode returns a NodeBuilder
+            // but we want to return a builder of this.type. This doesn't change anything as the return value is
+            // only a convenience for chaining
+            discard(builder.addFromVersionNode(node, nodes - nodeId))
+            builder
+          case None => throw NodeHashingError.IncompleteTransactionTree(nodeId)
+        }
+      }
+
+    def addNodesFromNodeIds(nodeIds: ImmArray[NodeId], nodes: Map[NodeId, Node]): this.type =
+      iterateOver(nodeIds)(addNodeFromNodeId(nodes))
```

Comment:
This will reinstall the full `nodes` map for each invocation. So the trick of removing the processed nodes does not work as described. You'll detect cycles in the tree this way, but you wouldn't be able to distinguish a DAG from a tree. Together with the exception `IncompleteTransactionTree` and its message, this check IMO does more harm than it buys us.

Reply:
I removed it